File: VB/PropertyGridWithDictionary/DictionaryWrapper.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.ComponentModel

Namespace PropertyGridWithDictionary
	Friend Class DictionaryWrapper(Of T)
		Implements ICustomTypeDescriptor

		Public Dict As Dictionary(Of String, T)

		Private _collection As PropertyDescriptorCollection

		Public Sub New(ByVal dict As Dictionary(Of String, T))
			Me.Dict = dict
			Dim coll As New List(Of PropertyDescriptor)()
			For Each key As String In Me.Dict.Keys
				coll.Add(New CustomPropertyDescriptor(Of T)(GetType(DictionaryWrapper(Of T)), key, Me.Dict(key).GetType(), New Attribute() {}))
			Next key
			_collection = New PropertyDescriptorCollection(coll.ToArray())
		End Sub

		Public Function GetAttributes() As AttributeCollection
			Return New AttributeCollection()
		End Function

		Public Function GetClassName() As String
			Return ""
		End Function

		Public Function GetComponentName() As String
			Return ""
		End Function

		Public Function GetConverter() As TypeConverter
			Return New TypeConverter()
		End Function

		Public Function GetDefaultEvent() As EventDescriptor
			Return Nothing
		End Function

		Public Function GetDefaultProperty() As PropertyDescriptor
			Return _collection(0)
		End Function

		Public Function GetEditor(ByVal editorBaseType As Type) As Object
			Return New Object()
		End Function

		Public Function GetEvents(ByVal attributes() As Attribute) As EventDescriptorCollection
			Return Nothing
		End Function

		Public Function GetEvents() As EventDescriptorCollection
			Return Nothing
		End Function

		Public Function GetProperties(ByVal attributes() As Attribute) As PropertyDescriptorCollection
			Return _collection
		End Function

		Public Function GetProperties() As PropertyDescriptorCollection
			Return _collection
		End Function

		Public Function GetPropertyOwner(ByVal pd As PropertyDescriptor) As Object
			Return Me
		End Function
	End Class
End Namespace

File: CS/PropertyGridWithDictionary/DictionaryWrapper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.ComponentModel;

namespace PropertyGridWithDictionary {
    class DictionaryWrapper<T> : ICustomTypeDescriptor {
        public Dictionary<string, T> Dict;

        private PropertyDescriptorCollection _collection;

        public DictionaryWrapper(Dictionary<string, T> dict) {
            Dict = dict;
            List<PropertyDescriptor> coll = new List<PropertyDescriptor>();
            foreach (string key in Dict.Keys)
                coll.Add(new CustomPropertyDescriptor<T>(typeof(DictionaryWrapper<T>), key, Dict[key].GetType(), new Attribute[] { }));
            _collection = new PropertyDescriptorCollection(coll.ToArray());
        }

        public AttributeCollection GetAttributes() {
            return new AttributeCollection();
        }

        public string GetClassName() {
            return "";
        }

        public string GetComponentName() {
            return "";
        }

        public TypeConverter GetConverter() {
            return new TypeConverter();
        }

        public EventDescriptor GetDefaultEvent() {
            return null;
        }

        public PropertyDescriptor GetDefaultProperty() {
            return _collection[0];
        }

        public object GetEditor(Type editorBaseType) {
            return new object();
        }

        public EventDescriptorCollection GetEvents(Attribute[] attributes) {
            return null;
        }

        public EventDescriptorCollection GetEvents() {
            return null;
        }

        public PropertyDescriptorCollection GetProperties(Attribute[] attributes) {
            return _collection;
        }

        public PropertyDescriptorCollection GetProperties() {
            return _collection;
        }

        public object GetPropertyOwner(PropertyDescriptor pd) {
            return this;
        }
    }
}